README.md
# AC AdobeIO
This is a SDK for the Adobe Stock footage API.

## Usage
```
const acAdobeIO = require('ac-adobeio')

acAdobeIO.init({
  clientId: 'xxx',
  clientSecret: 'xxx',
  jwtParams: {
    "iss": "xxx@AdobeOrg",
    "sub": "xxx@techacct.adobe.com",
    key: fs.readFileSync('private.key'),
  }
})

acAdobeIO.profile({}, (err, result) => {
  console.log(err, result)
})
```



CHANGELOG.md
<a name="1.1.1"></a>

## [1.1.1](https://github.com/AdmiralCloud/ac-adobeio/compare/v1.1.0..v1.1.1) (2023-06-26 17:50:51)


### Bug Fix

* **AdobeIO:** Bring version up-to-date | MP | [1e0e240a77f7c50fafe29f06109f84177672467e](https://github.com/AdmiralCloud/ac-adobeio/commit/1e0e240a77f7c50fafe29f06109f84177672467e)    
Update packages, remove unnecessary procedures  
Related issues: [/issues#undefined](https://github.com//issues/undefined)
### Chores

* **Misc:** Upddated packages | MP | [87e779c181828d189d9792e419c5d8a33141794d](https://github.com/AdmiralCloud/ac-adobeio/commit/87e779c181828d189d9792e419c5d8a33141794d)    
Updated packages
* **Misc:** Updated packages and use yarn | MP | [06107e824e9ad81deda67c3aa0313d4497826770](https://github.com/AdmiralCloud/ac-adobeio/commit/06107e824e9ad81deda67c3aa0313d4497826770)    
Updated packages and use yarn
<a name="1.1.0"></a>
# [1.1.0](https://github.com/mmpro/ac-adobeio/compare/v1.0.4...v1.1.0) (2019-06-17 17:44)


### Features

* **Misc:** Adobe API Update | Get ALL licenses | PL ([d7e4285](https://github.com/mmpro/ac-adobeio/commit/d7e4285))    
  https://www.adobe.io/apis/creativecloud/stock/docs.html#!adobe/stock-api-docs/master/docs/api/13-license-history.md



<a name="1.0.4"></a>
## [1.0.4](https://github.com/mmpro/ac-adobeio/compare/v1.0.3...v1.0.4) (2018-12-05 18:46)


### Bug Fixes

* **Misc:** Add search | MP ([9315614](https://github.com/mmpro/ac-adobeio/commit/9315614))    
  Add search (only for words at the moment)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/mmpro/ac-adobeio/compare/v1.0.2...v1.0.3) (2018-12-05 16:16)


### Bug Fixes

* **Misc:** Add info endpoint | MP ([3f2df91](https://github.com/mmpro/ac-adobeio/commit/3f2df91))    
  Add info endpoint and method



<a name="1.0.2"></a>
## [1.0.2](https://github.com/mmpro/ac-adobeio/compare/v1.0.1...v1.0.2) (2018-10-24 11:02)


### Bug Fixes

* **Misc:** LicenseHistory now support offset, limit and locale | MP ([77ec463](https://github.com/mmpro/ac-adobeio/commit/77ec463))    
  LicenseHistory now support offset, limit and locale



<a name="1.0.1"></a>
## [1.0.1](https://github.com/mmpro/ac-adobeio/compare/v1.0.0...v1.0.1) (2018-10-17 12:26)


### Bug Fixes

* **Misc:** Add method to return JWT token | MP ([1f25274](https://github.com/mmpro/ac-adobeio/commit/1f25274))    
  A new method returns the JWT token to the embedding application.



<a name="1.0.0"></a>
# 1.0.0 (2018-10-17 09:27)


### Features

* **Misc:** Initial release | MP ([44c98cf](https://github.com/mmpro/ac-adobeio/commit/44c98cf))    
  Initial release





test/test.js
/*
const expect = require('expect')
const acAdobeIO = require('../index')

describe('Testing TODO', () => {

})
*/


.czConfig.js
module.exports = {
  repository: {
    baseUrl: 'https://github.com/',
    repoUrl: 'mmpro/ac-adobeio/'
  },
  types: [
    {value: 'feat',     name: 'feat:     A new feature'},
    {value: 'fix',      name: 'fix:      A bug fix'},
    {value: 'docs',     name: 'docs:     Documentation only changes'},
    {value: 'style',    name: 'style:    Changes that do not affect the meaning of the code\n            (white-space, formatting, missing semi-colons, etc)'},
    {value: 'refactor', name: 'refactor: A code change that neither fixes a bug nor adds a feature'},
    {value: 'perf',     name: 'perf:     A code change that improves performance'},
    {value: 'test',     name: 'test:     Adding missing tests'},
    {value: 'chore',    name: 'chore:    Changes to the build process or auxiliary tools\n            and libraries such as documentation generation'},
    {value: 'revert',   name: 'revert:   Revert to a commit'},
    {value: 'WIP',      name: 'WIP:      Work in progress'}
  ],
  scopes: [
    {name: 'Misc' }
  ],
  allowCustomScopes: true,
  allowBreakingChanges: ['feat', 'fix']
};


index.js
const _ = require('lodash')
const request = require('superagent')
const jwt = require('jsonwebtoken')
const moment = require('moment')

const config = require('./config')

const acAdobeIO = () => {
  let clientId
  let clientSecret
  let product = 'AdmiralCloud Adobe Stock IO SDK'
  let jwtParams = {
    iss: 'xx',
    sub: 'xxx',
    key: 'xxx'
  }

  let jwtToken = {
    access_token: 'xxx',
    expires: 0
  }

  const init = (params) => {
    clientId = _.get(params, 'clientId')
    clientSecret = _.get(params, 'clientSecret')
    _.merge(jwtParams, _.get(params, 'jwtParams'))
    if (_.get(params, 'product')) product = _.get(params, 'product')
  }

  const getJWT = () => {
    return jwtToken
  }

  const requestJWT = (cb) => {
    // check if JWT is still valid
    if (_.get(jwtToken, 'expires') > moment().unix()) {
      return cb()
    }

    let tokenPayload = {
      'exp': moment().add(1, 'day').unix(),
      'iss': _.get(jwtParams, 'iss'),
      'sub': _.get(jwtParams, 'sub'),
      'https://ims-na1.adobelogin.com/s/ent_stocksearch_sdk': true,
      'aud': _.get(config, 'jwt.audPrefix') + clientId
    }
    let payload = {
      jwt_token: jwt.sign(tokenPayload, _.get(jwtParams, 'key'), { algorithm: 'RS256' }),
      client_id: clientId,
      client_secret: clientSecret
    }
    request.post(_.get(config, 'jwt.endpoint'))
      .set({
        'Content-Type': 'application/x-www-form-urlencoded'
      })
      .send(payload)
      .end((err, res) => {
        if (err) return cb(err)
        jwtToken = {
          access_token: _.get(res.body, 'access_token'),
          expires: _.get(tokenPayload, 'exp', 0)
        }
        return cb()
      })
  }

  /**
   *
   * @param params.contentId
   * @param params.license STRING Defaults to "Standard"
   * @param params.locale  STRING Defaults to "en_US"
   * @param {*} cb
   *
   *  { available_entitlement:
        { quota: 49,
          license_type_id: 42,
          has_credit_model: false,
          has_agency_model: true,
          is_cce: true,
          full_entitlement_quota: { standard_credits_quota: 49 } },
        member: { stock_id: 123456 },
        purchase_options:
        { state: 'possible',
          requires_checkout: false,
          message: 'This will use 1 of your 49 images.'
      }
    }
   *
   */

  const profile = (params, cb) => {
    requestJWT((err) => {
      if (err) return cb(err)

      const contentId = _.get(params, 'contentId')
      const license = _.get(params, 'license')
      const locale = _.get(params, 'locale')

      let payload = {
        content_id: contentId,
        license,
        locale
      }

      request
        .get(_.get(config, 'endpoints.profile'))
        .set({
          'x-api-key': clientId,
          'x-product': product,
          'Authorization': 'Bearer ' + _.get(jwtToken, 'access_token')
        })
        .query(payload)
        .end((err, res) => {
          return cb(err, _.get(res, 'body', {}))
        })
    })
  }

  /**
   * Returns the license history for the given user
   * https://www.adobe.io/apis/creativecloud/stock/docs.html#!adobe/stock-api-docs/master/docs/api/13-license-history.md
   *
   * @param params.locale
   * @param params.offset
   * @param params.limit
   *
   * @param {*} cb (err, result)
   *
   * { nb_results: 1,
files:
  [ { license: 'Standard',
      license_date: '10/17/18, 8:20 AM',
      download_url: 'https://stock.adobe.com/Download/DownloadFileDirectly/xxxx',
      id: 75950374,
      title: 'five kittens',
      creator_name: 'adyafoto',
      creator_id: 205216144,
      content_url: 'https://stock.adobe.com/Rest/images/five-kittens/75950374',
      media_type_id: 1,
      vector_type: null,
      content_type: 'image/jpeg',
      height: 1667,
      width: 2500,
      details_url: 'https://stock.adobe.com/75950374?as_channel=affiliate&as_source=api&as_content=xxx' } ] }
    */

  const licenseHistory = (params, cb) => {
    let queryString = []
    if (_.get(params, 'offset')) {
      queryString.push('search_parameters[offset]=' + _.get(params, 'offset'))
    }
    if (_.get(params, 'limit')) {
      queryString.push('search_parameters[limit]=' + _.get(params, 'limit'))
    }
    if (_.get(params, 'locale')) {
      queryString.push('locale]=' + _.get(params, 'locale'))
    }
    // Enforces all results https://www.adobe.io/apis/creativecloud/stock/docs.html#!adobe/stock-api-docs/master/docs/api/13-license-history.md
    queryString.push('all=true')

    requestJWT((err) => {
      if (err) return cb(err)
      request
        .get(_.get(config, 'endpoints.licenseHistory'))
        .query(_.join(queryString, '&'))
        .set({
          'x-api-key': clientId,
          'x-product': product,
          'Authorization': 'Bearer ' + _.get(jwtToken, 'access_token')
        })
        .end((err, res) => {
          return cb(err, _.get(res, 'body', {}))
        })
    })
  }

  /**
   * Buys an item/license
   * @param params.content_id INT REQUIRED content id to licenses
   * @param params.license STRING REQUIRED License type (e.g 'Standard')
   * @param {*} cb (err, result)
   */
  const license = (params, cb) => {
    const contentId = _.get(params, 'contentId')
    if (!contentId) return cb({ message: 'contentId_required' })
    const license = _.get(params, 'license', 'Standard')

    requestJWT((err) => {
      if (err) return cb(err)

      const payload = {
        content_id: contentId,
        license
      }

      request.get(_.get(config, 'endpoints.license'))
        .set({
          'x-api-key': clientId,
          'x-product': product,
          'Authorization': 'Bearer ' + _.get(jwtToken, 'access_token')
        })
        .query(payload)
        .end((err, res) => {
          return cb(err, _.get(res, 'body', {}))
        })
    })
  }

  /**
   * Gets info about the content id
   * @param params.content_id INT REQUIRED content id to licenses
   * @param {*} cb (err, result)
   */
  const info = (params, cb) => {
    const contentId = _.get(params, 'contentId')
    if (!contentId) return cb({ message: 'contentId_required' })

    requestJWT((err) => {
      if (err) return cb(err)

      const payload = {
        content_id: contentId
      }

      request.get(_.get(config, 'endpoints.info'))
        .set({
          'x-api-key': clientId,
          'x-product': product,
          'Authorization': 'Bearer ' + _.get(jwtToken, 'access_token')
        })
        .query(payload)
        .end((err, res) => {
          return cb(err, _.get(res, 'body', {}))
        })
    })
  }

  /**
   * Search Adobe Stock Footage library
   * @param params.words STRING String to search for
   * @param {*} cb (err, result) -> { nb_results: 0, files: [] }
   *
   * TODO: Add more parameters
   */
  const search = (params, cb) => {
    requestJWT((err) => {
      if (err) return cb(err)

      const payload = {
        'search_parameters[words]': params.words
      }

      request.get(_.get(config, 'endpoints.search'))
        .set({
          'x-api-key': clientId,
          'x-product': product,
          'Authorization': 'Bearer ' + _.get(jwtToken, 'access_token')
        })
        .query(payload)
        .end((err, res) => {
          return cb(err, _.get(res, 'body', {}))
        })
    })
  }

  return {
    init,
    getJWT,
    requestJWT,
    profile,
    info,
    search,
    license,
    licenseHistory
  }
}

module.exports = acAdobeIO()
